torch/amp/grad_scaling.py

```diff
@@ -0,0 +1,382 @@
+    not as a free function.
+
+    ``found_inf`` should be the result of an earlier call to ``optimizer.unscale`` called on the same optimizer
+    instance after all of its gradients have been populated.
+    By default, :func:`step_after_unscale` checks the value contained by ``found_inf``.  If ``found_inf`` contains 0.0
+    (in other words if the gradients do not contain infs/nans), :func:`step_after_unscale` runs the underlying
+    ``self.step()``.  If ``found_inf`` contains > 0, indicating that the gradients do contain infs or nans,
+    it skips the step().
+
+    Arguments:
+        found_inf (torch.cuda.FloatTensor):  1-element Tensor produced by an earlier call to ``optimizer.unscale()``
+            that should contain > 0 if the optimizer's current gradients have an inf/nan and 0.0 if they don't.
+        skip_if_inf (bool, optional, default=True):  If False, call ``step()`` without checking the value of
+            ``found_inf``.  If you're manually using the same gradient scale every iteration ("static gradient scaling"
+            as opposed to dynamic gradient scaling) and you're fairly
+            certain no inf/nan gradients will be encountered, setting ``skip_if_inf=False`` can provide a minor
+            performance improvement, because for typical optimizers it avoids copying ``found_inf``'s value back to
+            the CPU.  Another case where it makes sense to supply ``skip_if_inf=False`` is batch replay, where you
+            replay this iteration until inf/nan-free gradients are created.  By the time you reach the
+            ``optimizer.step_after_unscale()`` call, you're sure the gradients do not contain inf/nan, so it's safe
+            to elide the check.  For other (typical) use, the default ``skip_if_inf=True`` is strongly recommended.
+        scaling_enabled (bool, optional):  If ``False``, :func:`step_after_unscale` simply calls ``self.step()``
+            and returns ``opt_ret, None, None``, where opt_ret is the default return value of ``self.step()``.
+
+    Returns:
+        If ``found_inf`` contains 0.0, :func:`step_after_unscale` returns the default return value of the
+        ``self.step()`` call.  If ``found_inf`` contains > 0, :func:`step_after_unscale` returns None.
+    """
+    if not scaling_enabled:
+        return self.step(closure)
+
+    if closure is not None:
+        raise ValueError("Closure use is not currently supported.  It's tricky, but not impossible, and we're trying "
+                         "to decide if it's worth implementing.  If you require closure use, please comment on "
+                         "https://github.com/pytorch/pytorch/issues/25081, which will help us gauge demand.")
+
+    if skip_if_inf:
+        if found_inf.item():
+            return None
+        else:
+            return self.step()
+    else:
+        return self.step()
+
+
+def _add_amp_attributes(optimizer):
+    if not hasattr(optimizer, "unscale_and_step"):
```

Comment:
Should this function assert that the optimizer does not have these attributes? Otherwise it might define all of some of them with different semantics and the user would think they had patched the optimizer with AMP, which could be very confusing.